File: AlgorithmConcepts/BinarySearchTree/binarySearchTree.go
package main

import "fmt"

type node struct {
	value int
	left *node
	right *node
}

type Tree struct {
        root *node
        size int
}

func insertNode(value int, root node) {
     if root.value == 0 {
        root.value = value
     }
     if value < root.value {
        if root.left == nil {
           root.left = &node{value: value}
	} else {
           insertNode(value, *root.left)
	}
     } else {
        if root.right == nil {
           root.right = &node{value: value}
        } else {
           insertNode(value, *root.right)
        }		
     }
}

func searchNode(value int, root node) node {
     if value < root.value { 
        if root.left == nil {
           return node{}
        } else if root.left.value == value {
           return root
        } else {
           searchNode(value, *root.left)
        }
     } else if value > root.value{
        if root.right == nil {
           return node{}
        } else if root.right.value == value {
           return root
        } else {
           searchNode(value, *root.right)
        }
     } else {
         return root
     }
     return node{}
}

func deleteNode(value int) {

}

func minimum() {

}

func maximum() {

}

func displayTree() {

}

func main() {
     root := node{value: 3}
     Bst := Tree{root: &root}
     Bst.size++
     fmt.Println(Bst.root.value, Bst.root.left, Bst.root.right)
     insertNode(2,Bst.root)
     fmt.Println(Bst.root.value, Bst.root.left, Bst.root.right)
     //insertNode(4,root)
     //insertNode(5,root)
     var search_node = searchNode(2, root)
     fmt.Println(search_node)
}

File: AlgorithmConcepts/BinarySearchTree/binarySearchTree.py
# Nodes Structure

class Node:
	
	def __init__(self, value):
            self.value = value
            self.right = None
            self.left = None

	def print_node(self):
            print(self.value, self.right, self.left)


# Insert Node

# Logic: Traverse through the tree based on bst logic (left: lower, right: higher) and insert corresponding node

def insert(root, value):
    
      if value < root.value:
         if root.left == None:
            root.left = Node(value)
         else:
            insert(root.left, value)
      else:
         if root.right == None:
            root.right = Node(value)
         else:
            insert(root.right, value)
 

# Delete Node

# Find a node - search

def find(root, value):
    
    if root.value == value:
       return 1
    if value < root.value and root.left is not None:
       find(root.left, value)
    else:
       if root.right is not None:
       	find(root.right, value)
    return 0

# isBinary Search Tree

def isBST(root):
    
    if root.value > root.left.value and root.value < root.right.value:
       return 1
    else:
       return 0

    if root.left != None:
       isBST(root.left)
    if root.right != None:
       isBST(root.right)

# Breath first search

# Depth first search

# is a Balanced binary search tree

# Height of a binary search tree

# Maximum subtree 

# Maximum balanced subtree

# Print a binary search tree

def printBST(root):
    if root.value == None:
       return
    print(str(root.value))
    if root.left is not None:
       #print("\t\t"+"//")
       printBST(root.left,)
    if root.right is not None:
       #print("\t\t"+"\\")
       printBST(root.right,)
          

# Main Driver Program

def main():
    root = Node(3)
    insert(root,4)
    insert(root,2)
    insert(root,1)
    insert(root,3)
    printBST(root)	
    print find(root,3)
    print find(root,5)
    print isBST(root)
main()
